From a339870aa4f566cd26ef9c9fe89729b3c22f3afa Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Tue, 1 Jul 2014 17:07:38 -0400
Subject: [fs] nfs: nfs4_fl_prepare_ds, fix bugs when the connect attempt fails

Message-id: <1404234458-8484-1-git-send-email-steved@redhat.com>
Patchwork-id: 85158
O-Subject: [RHEL7.1 PATCH 1/1][BZ1113248] NFSv4.1: nfs4_fl_prepare_ds - fix bugs when the connect attempt fails
Bugzilla: 1113248
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Andrea Arcangeli <aarcange@redhat.com>

From: Trond Myklebust <Trond.Myklebust@netapp.com>

- Fix an Oops when nfs4_ds_connect() returns an error.
- Always check the device status after waiting for a connect to complete.

Reported-by: Andy Adamson <andros@netapp.com>
Reported-by: Jeff Layton <jlayton@redhat.com>
Signed-off-by: Trond Myklebust <Trond.Myklebust@netapp.com>
Cc: <stable@vger.kernel.org> # v3.10+
Upstream Id: 52b26a3e1bb3e065c32b3febdac1e1f117d88e15
Bugzilla: 1113248

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/filelayout/filelayoutdev.c b/fs/nfs/filelayout/filelayoutdev.c
index 5d94b22..7f8b417 100644
--- a/fs/nfs/filelayout/filelayoutdev.c
+++ b/fs/nfs/filelayout/filelayoutdev.c
@@ -804,9 +804,6 @@ nfs4_fl_prepare_ds(struct pnfs_layout_segment *lseg, u32 ds_idx)
  struct nfs4_deviceid_node *devid = FILELAYOUT_DEVID_NODE(lseg);
  struct nfs4_pnfs_ds *ret = ds;
 
- if (filelayout_test_devid_unavailable(devid))
-  return NULL;
-
  if (ds == NULL) {
   printk(KERN_ERR "NFS: %s: No data server for offset index %d\n",
    __func__, ds_idx);
@@ -815,7 +812,7 @@ nfs4_fl_prepare_ds(struct pnfs_layout_segment *lseg, u32 ds_idx)
  }
  smp_rmb();
  if (ds->ds_clp)
-  goto out;
+  goto out_test_devid;
 
  if (test_and_set_bit(NFS4DS_CONNECTING, &ds->ds_state) == 0) {
   struct nfs_server *s = NFS_SERVER(lseg->pls_layout->plh_inode);
@@ -829,6 +826,7 @@ nfs4_fl_prepare_ds(struct pnfs_layout_segment *lseg, u32 ds_idx)
   /* Either ds is connected, or ds is NULL */
   nfs4_wait_ds_connect(ds);
  }
+out_test_devid:
  if (filelayout_test_devid_unavailable(devid))
   ret = NULL;
 out:
-- 
1.7.1